File: 3640.trionic-array-ii.py
#
# @lc app=leetcode id=3640 lang=python3
#
# [3640] Trionic Array II
#

# @lc code=start
class Solution:
    def maxSumTrionic(self, nums: List[int]) -> int:
        n = len(nums)
        # inc1: max sum of the first increasing segment ending at current index
        # dec: max sum of (inc1 + decreasing segment) ending at current index
        # inc2: max sum of (inc1 + dec + increasing segment) ending at current index
        
        # Initialize with a very small number representing infinity
        inf = float('inf')
        inc1 = -inf
        dec = -inf
        inc2 = -inf
        
        ans = -inf
        
        for i in range(1, n):
            curr = nums[i]
            prev = nums[i-1]
            
            # Calculate new states based on transition from prev to curr
            # We use temporary variables for the next state to avoid using updated values within the same step
            if curr > prev:
                # Increasing step
                
                # update inc2: can extend previous inc2 OR start inc2 from a previous dec
                # Since we are going UP, we can be in the second increasing phase
                new_inc2 = -inf
                if inc2 != -inf:
                    new_inc2 = max(new_inc2, inc2 + curr)
                if dec != -inf:
                    new_inc2 = max(new_inc2, dec + curr)
                
                # update inc1: can extend previous inc1 OR start new inc1 with (prev, curr)
                new_inc1 = prev + curr
                if inc1 != -inf:
                    new_inc1 = max(new_inc1, inc1 + curr)
                
                # Cannot be in decreasing phase if we just went up
                new_dec = -inf
                
                inc1, dec, inc2 = new_inc1, new_dec, new_inc2
                
            elif curr < prev:
                # Decreasing step
                
                # update dec: can extend previous dec OR start dec from a previous inc1
                # Since we are going DOWN, we can be in the middle decreasing phase
                new_dec = -inf
                if dec != -inf:
                    new_dec = max(new_dec, dec + curr)
                if inc1 != -inf:
                    new_dec = max(new_dec, inc1 + curr)
                
                # Cannot extend increasing phases if we just went down
                new_inc1 = -inf
                new_inc2 = -inf
                
                inc1, dec, inc2 = new_inc1, new_dec, new_inc2
                
            else:
                # Equal elements: strictly increasing/decreasing property broken
                inc1, dec, inc2 = -inf, -inf, -inf
            
            if inc2 != -inf:
                ans = max(ans, inc2)
                
        return ans
# @lc code=end

File: 3640.trionic-array-ii.cpp
#
# @lc app=leetcode id=3640 lang=cpp
#
# [3640] Trionic Array II
#

# @lc code=start
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    long long maxSumTrionic(vector<int>& nums) {
        int n = nums.size();
        vector<long long> pre(n + 1, 0);
        for (int i = 1; i <= n; ++i) {
            pre[i] = pre[i - 1] + nums[i - 1];
        }
        const long long INF = LLONG_MAX / 2;
        const long long NINF = LLONG_MIN / 2;

        // Left increasing for p
        vector<int> L(n);
        vector<long long> gmin(n, INF);
        L[0] = 0;
        gmin[0] = INF;
        for (int p = 1; p < n; ++p) {
            if (nums[p] > nums[p - 1]) {
                L[p] = L[p - 1];
                gmin[p] = min(gmin[p - 1], pre[p - 1]);
            } else {
                L[p] = p;
                gmin[p] = INF;
            }
        }

        // Right increasing for q
        vector<int> R(n);
        vector<long long> fmax(n, NINF);
        R[n - 1] = n - 1;
        fmax[n - 1] = NINF;
        for (int q = n - 2; q >= 0; --q) {
            if (nums[q + 1] > nums[q]) {
                R[q] = R[q + 1];
                fmax[q] = max(pre[q + 2], fmax[q + 1]);
            } else {
                R[q] = q;
                fmax[q] = NINF;
            }
        }

        // Dec left for q
        vector<int> D(n);
        D[0] = 0;
        for (int i = 1; i < n; ++i) {
            if (nums[i] < nums[i - 1]) {
                D[i] = D[i - 1];
            } else {
                D[i] = i;
            }
        }

        // Sparse table for min on gmin
        int LOG = 0;
        while ((1 << LOG) <= n) ++LOG;
        vector<vector<long long>> st(LOG, vector<long long>(n));
        for (int i = 0; i < n; ++i) {
            st[0][i] = gmin[i];
        }
        for (int j = 1; j < LOG; ++j) {
            for (int i = 0; i + (1 << j) <= n; ++i) {
                st[j][i] = min(st[j - 1][i], st[j - 1][i + (1 << (j - 1))]);
            }
        }
        vector<int> logg(n + 1, 0);
        for (int i = 2; i <= n; ++i) {
            logg[i] = logg[i / 2] + 1;
        }

        auto query = [&](int le, int ri) -> long long {
            if (le > ri) return INF;
            int len = ri - le + 1;
            int k = logg[len];
            return min(st[k][le], st[k][ri - (1 << k) + 1]);
        };

        long long ans = NINF;
        for (int q = 0; q < n; ++q) {
            if (fmax[q] == NINF) continue;
            int dl = D[q];
            if (dl >= q) continue;
            long long min_g = query(dl, q - 1);
            if (min_g == INF) continue;
            long long cursum = fmax[q] - min_g;
            ans = max(ans, cursum);
        }
        return ans;
    }
};
# @lc code=end